sm8650.c

```diff
@@ -0,0 +1,468 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <sys/mman.h>
+#include <err.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "debugcc.h"
+
+static struct gcc_mux gcc = {
+	.mux = {
+		.phys =	0x100000,
+		.size = 0x1f4200,
+
+		.measure = measure_gcc,
+
+		.enable_reg = 0x62004,
+		.enable_mask = BIT(0),
+
+		.mux_reg = 0x62024,
+		.mux_mask = 0x3ff,
+
+		.div_reg = 0x62000,
+		.div_mask = 0xf,
+		.div_val = 2,
+	},
+
+	.xo_div4_reg = 0x62008,
+	.debug_ctl_reg = 0x62048,
+	.debug_status_reg = 0x6204c,
+};
+
+/* All leaf clocks are disabled for now untill we enable corresponding GDSCs */
+static struct debug_mux cam_cc = {
+	.phys = 0xade0000,
+	.size = 0x20000,
+	.block_name = "cam",
+
+	.measure = measure_leaf,
+	.parent = &gcc.mux,
+	.parent_mux_val = 0x6a,
+
+	.enable_reg = 0x14008,
+	.enable_mask = BIT(0),
+
+	.mux_reg = 0x16000,
+	.mux_mask = 0xff,
+
+	.div_reg = 0x14004,
+	.div_mask = 0xf,
+	.div_val = 4,
+};
+
+static struct debug_mux disp_cc = {
+	.phys = 0xaf00000,
+	.size = 0x20000,
+	.block_name = "disp",
+
+	.measure = measure_leaf,
+	.parent = &gcc.mux,
+	.parent_mux_val = 0x6f,
+
+	.enable_reg = 0xd004,
+	.enable_mask = BIT(0),
+
+	.mux_reg = 0x11000,
+	.mux_mask = 0x1ff,
+
+	.div_reg = 0xd000,
+	.div_mask = 0xf,
+	.div_val = 4,
+};
+
+static struct debug_mux gpu_cc = {
+	.phys = 0x3d90000,
+	.size = 0xa000,
+	.block_name = "gpu",
+
+	.measure = measure_leaf,
+	.parent = &gcc.mux,
+	.parent_mux_val = 0x18d,
+
+	.enable_reg = 0x9274,
+	.enable_mask = BIT(0),
+
+	.mux_reg = 0x9564,
+	.mux_mask = 0xff,
+
+	.div_reg = 0x9270,
+	.div_mask = 0xf,
+	.div_val = 2,
+};
+
+static struct debug_mux video_cc = {
+	.phys = 0xaaf0000,
+	.size = 0x10000,
+	.block_name = "video",
+
+	.measure = measure_leaf,
+	.parent = &gcc.mux,
+	.parent_mux_val = 0x78,
+
+	.enable_reg = 0x80fc,
+	.enable_mask = BIT(0),
+
+	.mux_reg = 0x9a4c,
+	.mux_mask = 0x3f,
+
+	.div_reg = 0x80f8,
+	.div_mask = 0xf,
+	.div_val = 3,
+};
+
+static struct measure_clk sm8650_clocks[] = {
+	/* gcc entries */
+	/*
+	 * GCC Debug Muxes:
+	 * - apss_cc_debug_mux: 0x135
+	 * - cam_cc_debug_mux: 0x6a
+	 * - disp_cc_debug_mux: 0x6f
+	 * - gpu_cc_debug_mux: 0x18d
+	 * - video_cc_debug_mux: 0x78
+	 */
```

Comment:
We can probably drop this comment, it is self-documented in the data structures above.